### ISSUE TITLE
Memory leak

### ISSUE THREAD

**Issue by MusicManiac** [dwesterwick/SPTLateToTheParty#1]
While running the mod, ram slowly and steadily creeps to 100%, eventually making game unplayable.

It continuously throws this error while in raid:
```
EXCEPTION: Newtonsoft.Json.JsonSerializationException: Error converting value {null} to type 'System.Double'. Path '62e281349ecd3f493f6df954.value', line 1, position 2481. ---> System.InvalidCastException: Null object cannot be converted to a value type.
```
which is SR-2M Zenit B-17 mount according to https://db.sp-tarkov.com/search/62e281349ecd3f493f6df954

here are my logs, you probably want to look at errors/traces.
[2023.07.09_20-23-27_0.13.0.5.23399 assetBundle.log](https://github.com/dwesterwick/SPTLateToTheParty/files/11996025/2023.07.09_20-23-27_0.13.0.5.23399.assetBundle.log)
[2023.07.09_20-23-27_0.13.0.5.23399 errors.log](https://github.com/dwesterwick/SPTLateToTheParty/files/11996026/2023.07.09_20-23-27_0.13.0.5.23399.errors.log)
[2023.07.09_20-23-27_0.13.0.5.23399 notifications.log](https://github.com/dwesterwick/SPTLateToTheParty/files/11996027/2023.07.09_20-23-27_0.13.0.5.23399.notifications.log)
[2023.07.09_20-23-27_0.13.0.5.23399 speaker.log](https://github.com/dwesterwick/SPTLateToTheParty/files/11996028/2023.07.09_20-23-27_0.13.0.5.23399.speaker.log)
[2023.07.09_20-23-27_0.13.0.5.23399 traces.log](https://github.com/dwesterwick/SPTLateToTheParty/files/11996029/2023.07.09_20-23-27_0.13.0.5.23399.traces.log)
[2023.07.09_20-23-27_0.13.0.5.23399_backend.log](https://github.com/dwesterwick/SPTLateToTheParty/files/11996030/2023.07.09_20-23-27_0.13.0.5.23399_backend.log)
[2023.07.09_20-23-27_0.13.0.5.23399 application.log](https://github.com/dwesterwick/SPTLateToTheParty/files/11996032/2023.07.09_20-23-27_0.13.0.5.23399.application.log)

I believe putting some null checks may solve the issue.

**Comment by dwesterwick** [comment]
Thanks for the report.

It looks like the look-ranking data wasn't generated correctly, and the client DLL is constantly asking for it because it's failing to parse that file. That means there's actually two issues I need to fix. 

Could you please share your _user\mods\DanW-LateToTheParty-x.x.x\db\lootRanking.json_ file too? That will be super helpful for troubleshooting what might be happening with generating that data.

In the meantime, I'll modify the code that requests that data so it gives up after failing to retrieve it. That should suppress the constant errors and alleviate the leak.

**Comment by MusicManiac** [comment]
Here's whole mod folder just in case
[DanW-LateToTheParty-1.2.0.zip](https://github.com/dwesterwick/SPTLateToTheParty/files/11996827/DanW-LateToTheParty-1.2.0.zip)

**Comment by dwesterwick** [comment]
Perfect, thanks! I see the issue in the data you shared, so I'll work on troubleshooting.

**Comment by dwesterwick** [comment]
Could you please test this pre-release to see if it fixes the problems you were having? [https://github.com/dwesterwick/SPTLateToTheParty/releases/tag/1.2.1-alpha1](url)

It looks like there might be another mod you're using that's corrupting item prices, and that's what was causing the loot-ranking data for my mod to become corrupted. Starting with the 1.2.1 pre-release, this issue should no longer happen even if the prices become corrupted. Instead, you'll see warning messages in the server console whenever it finds an item with this problem, and it will assume the price is 0 instead. 

As a second level of protection, even if the loot-ranking data becomes corrupted, the client should now recognize this and make an error appear in the server console to explain what happened. 

Hopefully this works for you!

**Comment by MusicManiac** [comment]
Nope. Tested on interchange. With mod - same issue, slowly creeps to 24 GB until RAM is full.
Without mod RAM usage is stable 14 GB.

[DanW-LateToTheParty-1.2.1.zip](https://github.com/dwesterwick/SPTLateToTheParty/files/12003185/DanW-LateToTheParty-1.2.1.zip)

Edit: funny thing though, I removed DLL and meant to remove server counterpart but looks like I forgot, only noticed cuz for the first time saw that some items were out of stock on traders xD
Well at least I'll get to use traders running out of items part of the mod, I guess :D

**Comment by dwesterwick** [comment]
Yeah, you should still be able to use the trader-stock changes, but without the DLL, items that you just unlock will be out-of-stock until the trader resets or you restart the server. That should rarely be an issue though I'm guessing.

As another interim solution, you should be able to set **destroy_loot_during_raid.loot_ranking.enabled=false** in *config.json*. That should make the mod ignore all of the loot-ranking data, and in theory that should also "fix" the memory leak. 

I'm surprised there are still null values after my changes, so I'm definitely missing something. I'll take another look.

To help me troubleshoot, could you please share:
1. Your server log from when you experienced the leak
2. A list of the other mods you're using (a screenshot is fine)

Thanks.

**Comment by MusicManiac** [comment]
~~In archive I've sent there's folder with logs.~~ I'll recreate the issue and get you fresh server logs in 12 to 16 hours. In regards of what mod might bork the prices my closest guess would be LotsOfLoot, Broker, Blacklist or SVM (that recently got some update for fence assortment, maybe it screws the data somehow). Nothing else should be interacting with item prices, I think.

UPD: attached 1 of the server logs, has a bunch of late to party warns. Will get fresh log tomorrow.

![image](https://github.com/dwesterwick/SPTLateToTheParty/assets/14209235/3f8f5db9-b0a8-4256-8d22-0961d8405b92)
![image](https://github.com/dwesterwick/SPTLateToTheParty/assets/14209235/2a35c90c-9f12-461c-8259-f05925f5676b)

[server-2023-07-10-21.log](https://github.com/dwesterwick/SPTLateToTheParty/files/12010071/server-2023-07-10-21.log)

**Comment by dwesterwick** [comment]
Thanks, a new server log will help, but the one you uploaded does help a bit.

It looks like all of the warnings occur when getting Fence's inventory, and all of the items it complains about have special characters at the end of their names (namely stars) or are sealed weapon cases. Do you know what mod might modify item names or clone items and add special characters to their names?

Also, I assume you're using SPT-AKI 3.5.8? That message was truncated from the server log.

**Comment by MusicManiac** [comment]
Oh, ye, it's Eye of a trader or Item info, one of them marks expensive items with stars. And ye, SPT 358.

**Comment by dwesterwick** [comment]
I guess that was just a coincidence. I just tested both and didn't get any errors. I'll wait until you send a new server log tomorrow and then I'll look through your mod list and see if any of them might be causing it.

**Comment by MusicManiac** [comment]
Launched the game, did some inventory stuff and went into 7 min customs raid, in raid memory started at 10gb or something, by the end was 24gb
[logs.zip](https://github.com/dwesterwick/SPTLateToTheParty/files/12014809/logs.zip)

**Comment by dwesterwick** [comment]
As a sanity check, could you please share your _Aki_Data\Server\database\templates\prices.json_ file too? I just want to make sure a mod is making the price changes dynamically and your install isn't corrupted. Either way, I still need to update my code so the memory leak doesn't occur, but I'm just wondering how to test it.

**Comment by MusicManiac** [comment]
[prices.zip](https://github.com/dwesterwick/SPTLateToTheParty/files/12017278/prices.zip)

**Comment by dwesterwick** [comment]
After some testing, I was able to reproduce this when using "The Blacklist" mod. It dynamically adds a bunch of records to the table populated using *prices.json*, but... it doesn't do a very good job with making sure the records have valid data. A lot of entries are null, undefined, or NaN (which is why my simple null check didn't work in my first 1.2.1 pre-release). I confirmed your *prices.json* file matches mine, so it's not corrupted. 

Hopefully this new pre-release will fix it: [https://github.com/dwesterwick/SPTLateToTheParty/releases/tag/1.2.1-alpha2](url). You'll get a ton of warnings in the server console whenever you refresh Fence's inventory due to my mod finding all of the junk records, so that will cause a small delay. Fortunately, everything seems to work fine in-raid. 

The remaining work I have to do before closing this issue is:
- Get confirmation from you that the leak is fixed in the Alpha 2 release
- Create an issue in the repo for "The Blacklist"
- Figure out why the memory leak was happening in the first place. I'd be fine with the mod spamming the console with errors, but there shouldn't be a massive memory leak like you experienced.

**Comment by MusicManiac** [comment]
~~Yep, looks like no leak anymore.
Here's log from factory run.
[Logs.zip](https://github.com/dwesterwick/SPTLateToTheParty/files/12030412/Logs.zip)~~

Upon further testing I'm not so sure anymore, just doesn't fill memory so fast?
Played Interchange 30 min raid with mod - starting RAM usage in-raid is 10, slowly fills up to 24 usage over the course of next 30 in-raid minutes
[with_mod_log_2023.07.12_16-54-45_0.13.0.5.23399.zip](https://github.com/dwesterwick/SPTLateToTheParty/files/12031031/with_mod_log_2023.07.12_16-54-45_0.13.0.5.23399.zip)
Without client side of mod - fills up to 20 GB over same time in same map. So 4GB difference. But I suspect that it might be me not taking exact same route as I did first time? Obviously there are multiple factors. Will test on other maps and report back the findings a bit later. Honestly I'm just not sure anymore cuz, yknow, BSG game very well optimized. /s
[without_mod_log_2023.07.12_18-21-52_0.13.0.5.23399.zip](https://github.com/dwesterwick/SPTLateToTheParty/files/12031356/without_mod_log_2023.07.12_18-21-52_0.13.0.5.23399.zip)

upd: all good, its just normal tarkov ram thing

**Comment by dwesterwick** [comment]
Good to hear it's working for you!

I tested Factory with and without the memory leak occurring, and there is a MASSIVE difference in memory consumption. The game was immediately unplayable for me due to the stuttering it caused, so I'm surprised you were able to put up with it for more than a couple minutes. I also tested Interchange a little, and I'm also starting at ~10.5GB. My memory wasn't climbing nearly as fast as yours (with the memory leak fixed), but I was also just leaving the game running with no bots and I'm only using a small fraction of the mods that you are. I'll have to see what it reaches at the end of a raid when I have time to actually play the game for real. 

I traced the leak down to the HTTP request and the subsequent failed parsing of the JSON response, but I wasn't able to figure out why it's happening. Instead I put in a band-aid fix that should only allow the exception to happen once. BSG would be proud. 

I'll probably try again tomorrow and then post an actual release either way. In the meantime, here's a new pre-release that will suppress all of the annoying warning messages when you refresh Fence's inventory: [https://github.com/dwesterwick/SPTLateToTheParty/releases/tag/1.2.1-alpha3](url)

I created an issue in the repo for "The Blacklist", but there hasn't been any response yet. 

Thanks for your help with testing!
